### ISSUE TITLE
Tree construction is based on algorithm with quadratic complexity

### ISSUE THREAD

**Issue by jhrcek** [turboMaCk/lazy-tree-with-zipper#30]
Summary: This is **not** an immediate issue for us, just an idea for addition of more efficient tree-constructing function.

The implementation of `Lazy.Tree.fromList` (https://github.com/turboMaCk/lazy-tree-with-zipper/blob/ad09ba6dec4672c1b3fd03922e049f2d5726dffb/src/Lazy/Tree.elm#L460-L467) has quadratic complexity.

Most of the time this is not an issue, because you rarely need to expand (force) the whole (sub)tree, so the complexity is spread over time as each node is being expanded. But there are usecases, where such expansion is needed (remember unwindAllDatapoints from pro-next?).

To illustrate the issue imagine a pathological example: a tree that looks like a long path of 1000 nodes with no branching (1 -> 2 -> 3 -> 4 -> ... -> 1000). Expanding each level does linear traversal of the whole list, so expanding the whole tree will need to traverse the list 1000 times IF the tree is fully expanded.

An idea how this could be improved is that you could have less general function for constructing the tree, where the "thunk" in each node wouldn't have linear complexity, but just logarithmic. Something like

```elm
fromListFaster : (item -> comparable) -> (item -> Maybe comparable) -> List item -> Forest item
fromListFaster getItemId getParentId  =
    let dictFromParentIdToListOfChildren :: Dict comparable (LList item) 
         -- The point: this dict ^ is just created ONCE with n*log n complexity;
         -- The benefit: function that is hidden behind the lazy thunk would just be `log n` lookup in this dict instead of linear traversal of the whole list
        -- forest roots would be special-cased here, as their parent is Nothing so they wouldn't be in the above dict
        forestRoots = ...
        lookupChildren item = Dict.get (getItemId item) dictFromParentIdToListOfChildren |> Maybe.withDefault EMPTY_LIST
    in 
    -- call some helper function using lookupChildren
```

Anyway this is just an idea. Feel free to close it if you don't think it would be worth it.

**Comment by turboMaCk** [comment]
I don't think it's possible to make this function perform better. What I think is possible is to make this function better in terms of complexity by clever usage of continuation passing. I've experimented with this in past. The advantage is that such implementation won't blow up the stack in any case as it hits tail call elimination. However in terms of performance it's actually much slower probably because it causes a lot of heap allocations for closures during csp.

Have you experimented with some alternative implementation yourself?

**Comment by jhrcek** [comment]
> I don't think it's possible to make this function perform better.

I'll dedicate an hour over the weekend to try to prove this wrong :laughing: 

> Have you experimented with some alternative implementation yourself?

Not yet, it's just a hunch. In practice you will almost never have such pathological trees, that would make it necessary to have more efficient implementation.

**Comment by turboMaCk** [comment]
I'll definitely accept the alternative implementation even if it has trade-off of just being more efficient / tail recursive while not performing better in runtime. My experimentation with CSP happend before I did the lazy variant so I was comparing strict implementation using simpler quadratic complexity without tail recursion with variant using CSP. The tradeoff was that CSP never blowed the stack but was significantly slower in each step of the recursion (more than twice). That's being said I might have implement it in some sub optimal way.

The other functions to watch for are fold functions. These are in my opinion even more problematic when used in practice because they need to evaluate whole tree. Anyway feel free to focus on just the construction, maybe we can port these ideas to these later.

**Comment by turboMaCk** [comment]
released in 3.2.0
